the-collab-lab/tcl-77-smart-shopping-list
# Mark item as purchased & update DB

## Description



This change adds the ability for a user to mark an item on their list as purchased. When marking the item as purchased the database will be updated with the current date as the new `dateLastPurchased` and increment the `totalPurchases` of the item. The list item checks if it has been more than 24 hours since an item was marked as purchased the item will uncheck and be able to be marked as purchased again. If a user tries to mark an item as purchased again in 24 hours it will prevent them from triggering it again.

## Related Issue


Closes #9 

## Acceptance Criteria



## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓    | 💯 Enhancement |
|  ✓    | ✨ New Feature |

## Updates

### Before


<img width="1512" alt="Screenshot 2024-09-05 at 11 46 07 AM" src="https://github.com/user-attachments/assets/29461513-396a-4428-87af-82d93b3bd0c7">

### After


<img width="1512" alt="Screenshot 2024-09-06 at 1 59 11 PM" src="https://github.com/user-attachments/assets/f0c5e719-4ed3-42eb-9b03-809132116eba">


https://github.com/user-attachments/assets/67823bd4-405a-4a11-9fe0-9f1ff97d6eaf


## Testing Steps / QA Criteria


- From your terminal, pull down this branch with `git pull origin bb-ma/mark-item-purchased and check that branch out with git checkout `bb-ma/mark-item-purchased`
- Then `npm ci` to install the newly added dependencies locally and `npm start` to launch the app.
- Click one of your lists that has items in it.
- Click the list navigate link in the nav bar.
- All items in the list should now have a checkbox to the left of them.
- Click either the label or the checkbox to mark an item as purchased.
- Click again and you will get a notification that will say it can't be marked again in a 24 hour period. 
- Open a new tab with the [firebase console](https://console.firebase.google.com/u/0/project/tcl-77-smart-shopping-list/firestore/databases/-default-/data/~2F5mLi0TYfwKVk9BnM3fdRRR9BaYe2~2FCostco) and navigate to your lists (the unique id associated with you outside of the users collection).
- Open the list you just marked items purchased on.
- See that the `dateLastPurchased` and `totalPurchases` has been updated.
- Change the `dateLastPurchased` to be at least 24 hours and 1 minute in the past from the current `dateLastPurchased`.
- Return to the app and notice the checkbox should be unchecked and able to be updated again.
